### PR TITLE
Parallelize requests

### LINKED ISSUE #100: Parallelize queries to improve speed
`stats` script should end before running `license`, `feature` and `option`.

There's a parameter in config.json to choose to run `stats`:
if it's disabled, we can run everything immediately.

### PR DESCRIPTION
**Pull request by cirelli94**
Fixes #100

### CODE REVIEW

**Comment by cirelli94** [comment]
There's still some margin of optimization, but at the cost of more awful reading of the code.

File: builder/builder.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+	"github.com/ercole-io/ercole-agent/model"
+	"github.com/ercole-io/ercole-agent/utils"
+)
+
+func BuildData(configuration config.Configuration, version string, hostDataSchemaVersion int) *model.HostData {
+	hostData := new(model.HostData)
+
+	if configuration.Hostname != "default" {
+		hostData.Hostname = configuration.Hostname
+	}
```
Comment (amreo) [inline builder/builder.go]: Are you sure that the hostdata.Hostname should be checked here and not after the line 29?

File: builder/builder.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+	hostData.Environment = configuration.Envtype
+	hostData.Location = configuration.Location
+	hostData.HostType = configuration.HostType
+	hostData.Version = version
+	hostData.HostDataSchemaVersion = hostDataSchemaVersion
+	hostData.Info = *getHost(configuration)
+	hostData.Hostname = hostData.Info.Hostname
```
Comment (amreo) [inline builder/builder.go]: You are losing what you setted in the lines 19-21

File: builder/builder.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	}
+
+	return database
+}
+
+func getOpenDatabase(configuration config.Configuration, entry model.OratabEntry, hostType string) model.Database {
+	dbVersionOut := fetcher(configuration, "dbversion", entry.DBName, entry.OracleHome)
+	dbVersion := strings.Split(string(dbVersionOut), ".")[0]
+
+	statsCtx, cancelStatsCtx := context.WithCancel(context.Background())
```
Comment (amreo) [inline builder/builder.go]: Are you sure that context.Background is correct?
[godoc](https://golang.org/pkg/context/#TODO) says that it's never cancelled

File: builder/fetcher.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err = cmd.Run()
+	if len(stderr.Bytes()) > 0 {
+		log.Print(string(stderr.Bytes()))
+	}
+
+	if err != nil {
+		if fetcherName != "dbstatus" {
+			log.Fatal(err)
```
Comment (amreo) [inline builder/fetcher.go]: What happen when a gorouting panic()?

File: utils/routine.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	done := make(chan bool, 1)
+
+	go func(done chan<- bool) {
+		function()
+		done <- true
+	}(done)
+
+	if !configuration.ParallelizeRequests {
+		<-done
+	}
```
Comment (amreo) [inline utils/routine.go]: Why do you the done channel and don't simply call function() if !configuration.ParallelizeRequest?

**Comment by cirelli94** [comment]
> Have you already check that it works?

No, they are still recovering the vms after the breaking of the disk..

File: builder/builder.go (hunk 1 above)
Comment (cirelli94) [reply builder/builder.go]: You are surely right! I'm gonna fix this.

File: builder/builder.go (hunk 3 above)
Comment (cirelli94) [reply builder/builder.go]: context.Background() is the parent context, WithCancel returns a copy of parent with a new Done channel
https://golang.org/pkg/context/#WithCancel

File: builder/fetcher.go (hunk 4 above)
Comment (cirelli94) [reply builder/fetcher.go]: If a goroutine panic, the execution will stall.
This is surely a side effect we should avoid managing panic.
I'll have to manage this with errgroups https://godoc.org/golang.org/x/sync/errgroup

Comment (cirelli94) [reply builder/fetcher.go]: Or maybe we should avoid to panic and manage errors: it's easier and prettier.

Comment (amreo) [reply builder/fetcher.go]: Mhmhmh

Comment (cirelli94) [reply builder/fetcher.go]: `os.Exit` cause the entire process to exit, so `log.Fatal` is not a problem.

File: utils/routine.go (hunk 5 above)
Comment (cirelli94) [reply utils/routine.go]: Fixed! Thanks!

File: builder/fetcher.go (hunk 4 above)
Comment (cirelli94) [reply builder/fetcher.go]: https://play.golang.org/p/rwofMzlOLKY
Both panic and log.Fatal terminate the program.
If we use panic we can recover and manage the error inside the goroutine, but if before wasn't managed I don't think I should did it here.

**Pull request merged**